File: ios/Sources/BesitosOfferwall/UrlBuilder.swift
import Foundation

struct UrlBuilder {
    private static let baseUrl = "https://wall.besitos.ai/v1"

    static func build(config: OfferwallConfig) -> URL? {
        var components = URLComponents(string: "\(baseUrl)/\(encode(config.partnerId))/offers")
        var items: [URLQueryItem] = []

        items.append(URLQueryItem(name: "userid", value: config.userId))

        if let subId = config.subId, !subId.isEmpty {
            items.append(URLQueryItem(name: "subid", value: subId))
        }
        if let deviceId = config.deviceId, !deviceId.isEmpty {
            items.append(URLQueryItem(name: "device_id", value: deviceId))
        }
        if let idfa = config.idfa, !idfa.isEmpty {
            items.append(URLQueryItem(name: "idfa", value: idfa))
        }
        if let gdfa = config.gdfa, !gdfa.isEmpty {
            items.append(URLQueryItem(name: "gdfa", value: gdfa))
        }
        if let info = config.info, !info.isEmpty {
            items.append(URLQueryItem(name: "info", value: info))
        }
        if config.hideHeader {
            items.append(URLQueryItem(name: "hide_header", value: "1"))
        }
        if config.hideFooter {
            items.append(URLQueryItem(name: "hide_footer", value: "1"))
        }

        components?.queryItems = items
        return components?.url
    }

    private static func encode(_ value: String) -> String {
        value.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? value
    }
}

File: android/besitos-offerwall/src/main/kotlin/ai/besitos/offerwall/url/UrlBuilder.kt
package ai.besitos.offerwall.url

import ai.besitos.offerwall.SdkConfig
import ai.besitos.offerwall.config.OfferwallConfig
import java.net.URLEncoder

internal object UrlBuilder {

    fun build(config: OfferwallConfig): String {
        val params = buildList {
            add("userid" to config.userId)
            config.subId?.takeIf { it.isNotBlank() }?.let { add("subid" to it) }
            config.deviceId?.takeIf { it.isNotBlank() }?.let { add("device_id" to it) }
            config.idfa?.takeIf { it.isNotBlank() }?.let { add("idfa" to it) }
            config.gdfa?.takeIf { it.isNotBlank() }?.let { add("gdfa" to it) }
            config.info?.takeIf { it.isNotBlank() }?.let { add("info" to it) }
            if (config.hideHeader) add("hide_header" to "1")
            if (config.hideFooter) add("hide_footer" to "1")
        }

        val query = params.joinToString("&") { (k, v) ->
            "${encode(k)}=${encode(v)}"
        }

        return "${SdkConfig.BASE_URL}/${encode(config.partnerId)}/offers?$query"
    }

    private fun encode(value: String): String =
        URLEncoder.encode(value, "UTF-8").replace("+", "%20")
}
